bitcoin/src/blockdata/script/asm.rs

```diff
@@ -0,0 +1,160 @@
+                    builder = builder.push_slice(push);
+                } else {
+                    builder = builder.push_opcode(op);
+                }
+                continue;
+            }
+
+            // Not an opcode, try to interpret as number or push.
+
+            if word.starts_with('<') && word.ends_with('>') {
+                word = &word[1..word.len() - 1];
+            }
+
+            // Try a number.
+            if let Ok(i) = &word.parse::<i32>() {
+                builder = builder
+                    .push_int(*i)
+                    .map_err(|_| err(pos, AsmParseErrorKind::PushExceedsMaxSize))?;
+                continue;
+            }
+
+            // Finally, try hex in various forms.
+            if word.starts_with("0x") {
+                word = &word[2..];
+            }
+
+            if try_parse_raw_hex(word, &mut buf) {
+                let push = <&PushBytes>::try_from(&buf[..])
+                    .map_err(|_| err(pos, AsmParseErrorKind::PushExceedsMaxSize))?;
+                builder = builder.push_slice(push);
+            } else {
+                return Err(err(pos, AsmParseErrorKind::UnknownInstruction));
+            }
+        }
+
+        Ok(builder.into_script())
+    }
+}
+
+/// Creates an iterator over instruction words and their position in the file.
+fn iter_words(asm: &str) -> impl Iterator<Item = ((usize, usize), &str)> {
+    asm.lines().enumerate().flat_map(|(line_idx, line)| {
+        let content = line
+            .split("#")
+            .next()
+            .expect("Could not split on #")
+            .split("//")
+            .next()
+            .expect("Could not split on //");
+        content
+            .split_whitespace()
+            .enumerate()
+            .map(move |(word_idx, word)| ((line_idx, word_idx), word))
+    })
+}
+
+/// Try to parse raw hex bytes and push them into the buffer.
+fn try_parse_raw_hex(hex: &str, buf: &mut Vec<u8>) -> bool {
+    buf.clear();
+    let iter = match hex::HexToBytesIter::new(hex) {
+        Ok(i) => i,
+        Err(_) => return false,
+    };
+    for item in iter {
+        let item = match item {
+            Ok(i) => i,
+            Err(_) => return false,
+        };
+        buf.push(item);
+    }
+    true
+}
+
+fn err(position: (usize, usize), kind: AsmParseErrorKind) -> ParseAsmError {
+    ParseAsmError { position, kind }
+}
+
+/// Error from parsing Script ASM.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseAsmError {
+    /// The position of the instruction that caused the error.
+    ///
+    /// The value is (line, word) with word incremented after
+    /// every chunk of whitespace.
+    pub position: (usize, usize),
+
+    /// The kind of error that occurred.
+    pub kind: AsmParseErrorKind,
+}
+
+/// The different kinds of [`ParseAsmError`]s that can occur.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum AsmParseErrorKind {
+    /// ASM ended unexpectedly.
+    UnexpectedEOF,
```

Comment:
```suggestion
    UnexpectedEof,
```

Reply:
Thanks! Done in 9f16317109e7a2530d74267140afc0984ff330d0